### ISSUE TITLE
[Feature Request] Better refactoring

### ISSUE THREAD

**Issue by storm1er** [DEVSENSE/phptools-docs#213]
Hi,

I just tried vscode `marsl.vscode-php-refactoring` extension and it's really awesome.
I'm pretty sure DevSense could implement it and be better without a doubt given the current devsense extension has more information =)

Thanks

**Comment by jakubmisek** [comment]
Thank you for the suggestion, and the nice words!

As far as I see, the only additional feature is the "Extract Method" right?

**Comment by jakubmisek** [comment]
@ctf0 Thank you. Currently, the PHP editor has the https://docs.devsense.com/en/vscode/editor/refactoring and https://docs.devsense.com/en/vscode/editor/code-actions.

We have the "extract method" and more "code actions" on our to-do list.

**Comment by frees11** [comment]
Hi, do you have any ETA, when we could expect Extract Method feature?

**Comment by jakubmisek** [comment]
> Hi, do you have any ETA, when we could expect Extract Method feature?

No specific ETA yet, we'll be first doing some code actions. Next month would be more probable.

**Comment by frees11** [comment]
Hello any progress?

**Comment by jakubmisek** [comment]
@frees11 we're adding more code actions. "Extract Method" is being tested so we can release it in upcoming weeks.

**Comment by frees11** [comment]
@jakubmisek  WOW, great to see this. If you need beta tester I will be glad to participate.

**Comment by frees11** [comment]
@jakubmisek hi any progress in this matter?

**Comment by jakubmisek** [comment]
@frees11 we have added several new code actions and code fixes,
and internally we have the extract method implemented as well - now we're deciding on how to approach the UI of the Extract Method.

**Comment by frees11** [comment]
in my point of view, best UI for this has phpstorm - you will just select lines of code - click on extract to method, it will ask for new function name and thats it

**Comment by jakubmisek** [comment]
@frees11 I agree, we just don't have this UI in VS Code (asking for a new function name), so we'll be inserting a snippet with a placeholder for the new function name instead I guess.

**Comment by ctf0** [comment]
check the popup api to get the user input for anything

**Comment by frees11** [comment]
@jakubmisek hi any progress in this matter?

**Comment by jakubmisek** [comment]
thank you for reminding me, @frees11

**Comment by jakubmisek** [comment]
`extract method` is coming to the pre-release

**Comment by jakubmisek** [comment]
list of code actions (incl. extract function) can be found on https://docs.devsense.com/vscode/code%20actions/list/
